framework/db/SqlToken.php

```diff
@@ -0,0 +1,296 @@
+     */
+    public function getIsCollection()
+    {
+        return in_array($this->type, [
+            self::TYPE_CODE,
+            self::TYPE_STATEMENT,
+            self::TYPE_PARENTHESIS,
+        ], true);
+    }
+
+    /**
+     * Returns whether the token represents a collection of tokens and has non-zero number of children.
+     * @return bool whether the token has children.
+     */
+    public function getHasChildren()
+    {
+        return $this->getIsCollection() && !empty($this->_children);
+    }
+
+    /**
+     * Returns the SQL code representing the token.
+     * @return string SQL code.
+     */
+    public function getSql()
+    {
+        $code = $this;
+        while ($code->parent !== null) {
+            $code = $code->parent;
+        }
+        return mb_substr($code->content, $this->startOffset, $this->endOffset - $this->startOffset, 'UTF-8');
+    }
+
+    /**
+     * @param string $pattern
+     * @param int $offset
+     * @param int|null $firstMatchIndex
+     * @param int|null $lastMatchIndex
+     * @return bool
+     */
+    public function matches($pattern, $offset = 0, &$firstMatchIndex = null, &$lastMatchIndex = null)
+    {
+        $patternToken = (new \yii\db\sqlite\SqlTokenizer($pattern))->tokenize();
+        if (!$patternToken->getHasChildren()) {
+            return false;
+        }
+
+        $patternToken = $patternToken[0];
+        return $this->tokensMatch($patternToken, $this, $offset, $firstMatchIndex, $lastMatchIndex);
+    }
+
+    /**
+     * Tests the given token to match the specified pattern token.
+     * @param SqlToken $patternToken
+     * @param SqlToken $token
+     * @param int $offset
+     * @param int|null $firstMatchIndex
+     * @param int|null $lastMatchIndex
+     * @return bool
+     */
+    private function tokensMatch(SqlToken $patternToken, SqlToken $token, $offset = 0, &$firstMatchIndex = null, &$lastMatchIndex = null)
+    {
+        if (
+            $patternToken->getIsCollection() !== $token->getIsCollection()
+            || (!$patternToken->getIsCollection() && $patternToken->content !== $token->content)
+        ) {
+            return false;
+        }
+
+        if ($patternToken->children === $token->children) {
+            $firstMatchIndex = $lastMatchIndex = $offset;
+            return true;
+        }
+
+        $firstMatchIndex = $lastMatchIndex = null;
+        $wildcard = false;
+        for ($index = 0, $count = count($patternToken->children); $index < $count; $index++) {
+            if ($patternToken[$index]->content === 'any') {
+                $wildcard = true;
+                continue;
+            }
+
+            for ($limit = $wildcard ? count($token->children) : $offset + 1; $offset < $limit; $offset++) {
+                if (!$wildcard && !isset($token[$offset])) {
+                    break;
+                }
+
+                if (!$this->tokensMatch($patternToken[$index], $token[$offset])) {
```

Comment:
Same here. You're incrementing children index but accessing parent with it. Is it intended?